pkg/cmd/template/cmd_assert_validate_test.go

```diff
@@ -0,0 +1,273 @@
+  my_array:
+  #@assert/validate ("a non-empty string", lambda v: True if len(v) > 0 else assert.fail("length of string was 0"))
+  - abc
+  #@assert/validate ("an int over 9000", lambda v: True if v > 9000 else assert.fail("int was less than 9000"))
+  - 12345
+  #@assert/validate ("a float less than pi", lambda v: True if v < 3.1415 else assert.fail("float was more than 3.1415"))
+  - 3.14
+  #@assert/validate ("bool evaluating to true", lambda v:  v)
+  - true
+  #@assert/validate ("a null value", lambda v: True if v == None else assert.fail("value was not null"))
+  - null
+`
+
+		expected := `my_map:
+  string: server.example.com
+  int: 54321
+  float: 2.3
+  bool: true
+  nil: null
+  my_array:
+  - abc
+  - 12345
+  - 3.14
+  - true
+  - null
+`
+
+		filesToProcess := files.NewSortedFiles([]*files.File{
+			files.MustNewFileFromSource(files.NewBytesSource("schema.yml", []byte(dataValuesYAML))),
+		})
+
+		assertSucceedsDocSet(t, filesToProcess, expected, opts)
+	})
+
+	t.Run("assert fails with only data values files", func(t *testing.T) {
+		t.Run("on data values", func(t *testing.T) {
+			opts := cmdtpl.NewOptions()
+			dataValuesYAML := `#@ load("@ytt:assert", "assert")
+
+#@data/values
+---
+#@assert/validate ("a map with less than 3 elements", lambda v: True if len(v) < 3 else assert.fail("length of map was more than or equal to 3"))
+my_map:
+  #@assert/validate ("a non-empty string", lambda v: True if len(v) > 0 else assert.fail("length of string was 0"))
+  string: ""
+  #@assert/validate ("an int over 9000", lambda v: True if v > 9000 else assert.fail("int was less than 9000"))
+  int: 2
+  #@assert/validate ("a float less than pi", lambda v: True if v < 3.1415 else assert.fail("float was more than 3.1415"))
+  float: 20.3
+  #@assert/validate ("bool evaluating to true", lambda v:  v)
+  bool: false
+  #@assert/validate ("a null value", lambda v: True if v == None else assert.fail("value was not null"))
+  nil: anything else
+  #@assert/validate ("an array with more than or exactly 10 items", lambda v: True if len(v) >= 10 else assert.fail("array was less than 10 items"))
+  my_array:
+  #@assert/validate ("a non-empty string", lambda v: True if len(v) > 0 else assert.fail("length of string was 0"))
+  - ""
+  #@assert/validate ("an int over 9000", lambda v: True if v > 9000 else assert.fail("int was less than 9000"))
+  - 123
+  #@assert/validate ("a float less than pi", lambda v: True if v < 3.1415 else assert.fail("float was more than 3.1415"))
+  - 3.14159
+  #@assert/validate ("bool evaluating to true", lambda v:  v)
+  - false
+  #@assert/validate ("a null value", lambda v: True if v == None else assert.fail("value was not null"))
+  - not null
+`
+
+			expectedErr := `One or more data values were invalid:
+- schema.yml:6 requires "a map with less than 3 elements"; assert.fail: fail: length of map was more than or equal to 3 (by schema.yml:5)
```

Comment:
```suggestion
- my_map (at schema.yml:6) requires "a map with less than 3 elements"; assert.fail: fail: length of map was more than or equal to 3 (by schema.yml:5)
```

(see suggestion at `Rule.Validate()` for a possible approach)